### PR TITLE
Validator and metrics for non-standard date-time display formats

### PR DESCRIPTION
**Pull request by labkey-adam**
#### Rationale
First part of Date/Time formatting simplifications story: https://docs.google.com/document/d/1XcatQe9wRDw1ONLML9x4fFv_fIq9zhXPHoYANBuOykw/edit

Adds a validation provider for non-standard date display formats and a metrics provider that collects & reports all non-standard patterns.

#### Changes
- Define server-side lists of standard date and time format patterns. Add `isStandard*DisplayFormat()` methods. Push standard display formats into client-side `LABKEY` object.
- Rework site validation to register provider factories instead of providers. Creating a new provider on every run allows it to query and compute warnings before container iteration, which allows for a more efficient validation process in some cases.
- Add root container when validating "all projects & folders"
- Fix `ModuleReportResource` date parsing... don't use display formats to parse!
- Add validation for root/folder/project default date/time display formats
- Add validation for custom query and table override metadata (QueryDef XML)
- Add validation for provisioned table columns
- Add metrics provider implemented using the same code paths

### CODE REVIEW

File: core/src/org/labkey/core/admin/DisplayFormatAnalyzer.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+            ));
+        }
+
+        // First, inspect QueryDefinition to identify columns where XML has explicitly set a display format
+        // (as opposed to inheriting a display format from another query or table definition). Then inspect
+        // those columns via the TableInfo to determine date-time columns with non-standard formats.
+        QueryService qs = QueryService.get();
+        List<QueryException> errors = new ArrayList<>();
+        QueryUrls urls = PageFlowUtil.urlProvider(QueryUrls.class);
+        _queryCandidatesMap.get(c).stream()
```
Comment (labkey-susanh) [inline core/src/org/labkey/core/admin/DisplayFormatAnalyzer.java]: Seems like either the null check above for `_defaulFormatsMap` is not needed or it is needed here and below for `_propertyCandidatesMap`.

**Review by labkey-susanh — APPROVED** [review]

File: core/src/org/labkey/core/admin/DisplayFormatAnalyzer.java (hunk 1 above)
Comment (labkey-adam) [reply core/src/org/labkey/core/admin/DisplayFormatAnalyzer.java]: Indeed... `MultiValuedMap.get()` returns an empty collection for non-existent keys, so null check was dumb. Revised and simplified.

**Comment by labkey-matthewb** [comment]
Suggestion for the formatting of the messages.  The "more" link is useful for context, but maybe display the folder path consistently in the message?  Also, maybe some formatting for the actual format string to make it stand out.  e.g.

Instead of
Project default display format for DateTimes: hh 'o''clock' a, zzzz [More info](http://localhost:8080/home/admin-projectSettings.view)
                    Date property "lists.asdf.F": hh 'o''clock' a, zzzz [More info](http://localhost:8080/home/mystudy/query-begin.view?schemaName=lists&queryName=asdf)

Maybe
Default display format for DateTimes in Project /home : _hh 'o''clock' a, zzzz_ [More info](http://localhost:8080/home/admin-projectSettings.view)
                    Date format for column "lists.asdf.F" in Folder /home/mystudy : _hh 'o''clock' a, zzzz_ [More info](http://localhost:8080/home/mystudy/query-begin.view?schemaName=lists&queryName=asdf)

I see that they are actually grouped by folder.  So that's nice.

**Comment by labkey-matthewb** [comment]
Adding a non-standard format using _edit metadata_ in the schema browser did not show up in the site validation report.

E.g. setting the format for list.xyz.created to yyMMddHHmmssZ

File: core/src/org/labkey/core/admin/DisplayFormatAnalyzer.java [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+            ));
+
+        // First, inspect QueryDefinition to identify columns where XML has explicitly set a display format
+        // (as opposed to inheriting a display format from another query or table definition). Then inspect
+        // those columns via the TableInfo to determine date-time columns with non-standard formats.
+        List<QueryException> errors = new ArrayList<>();
+        _queryCandidatesMap.get(c).stream()
+            .map(candidate -> _queryService.getQueryDef(user, c, candidate.schemaName, candidate.queryName))
+            .forEach(definition -> {
+                TablesDocument doc = definition.getMetadataTablesDocument();
```
Comment (labkey-matthewb) [inline core/src/org/labkey/core/admin/DisplayFormatAnalyzer.java]: I hit an NPE here (definition==null), but it didn't repro.  Maybe needs .filter(Objects::nonNull) for safety?

**Comment by labkey-adam** [comment]
> Suggestion for the formatting of the messages. The "more" link is useful for context, but maybe display the folder path consistently in the message? Also, maybe some formatting for the actual format string to make it stand out. e.g.
> 
> Instead of Project default display format for DateTimes: hh 'o''clock' a, zzzz [More info](http://localhost:8080/home/admin-projectSettings.view) Date property "lists.asdf.F": hh 'o''clock' a, zzzz [More info](http://localhost:8080/home/mystudy/query-begin.view?schemaName=lists&queryName=asdf)
> 
> Maybe Default display format for DateTimes in Project /home : _hh 'o''clock' a, zzzz_ [More info](http://localhost:8080/home/admin-projectSettings.view) Date format for column "lists.asdf.F" in Folder /home/mystudy : _hh 'o''clock' a, zzzz_ [More info](http://localhost:8080/home/mystudy/query-begin.view?schemaName=lists&queryName=asdf)
> 
> I see that they are actually grouped by folder. So that's nice.

Yes, they are grouped by folder and project + folder name is listed immediately above each group, so seems redundant to include in every message as well.

Site validation messages have never supported HTML... but now they do. See latest commit. Format patterns are now rendered in italics.

File: core/src/org/labkey/core/admin/DisplayFormatAnalyzer.java (hunk 2 above)
Comment (labkey-adam) [reply core/src/org/labkey/core/admin/DisplayFormatAnalyzer.java]: Added `nonNull`

**Comment by labkey-adam** [comment]
> Adding a non-standard format using _edit metadata_ in the schema browser did not show up in the site validation report.
> 
> E.g. setting the format for list.xyz.created to yyMMddHHmmssZ

I'll take a look at this...

**Comment by labkey-adam** [comment]
> > Adding a non-standard format using _edit metadata_ in the schema browser did not show up in the site validation report.
> > E.g. setting the format for list.xyz.created to yyMMddHHmmssZ
> 
> I'll take a look at this...

This should be fixed... please try again

**Review by labkey-matthewb — APPROVED** [review]

**Pull request merged**